import "goby/common/protobuf/liaison_config.proto";
import "goby/acomms/protobuf/network_ack.proto";

package goby.common.protobuf;

extend goby.common.protobuf.LiaisonConfig
{
  optional MOOSScopeConfig moos_scope_config = 1000;
  optional ProtobufCommanderConfig pb_commander_config = 1001;
  optional GeodesyConfig geodesy_config = 1002;
  optional AcommsConfig acomms_config = 1003;
}

message ProtobufCommanderConfig
{
  repeated string load_protobuf_name = 1;
  
  optional int32 value_width_pixels = 10 [default = 500];
  optional int32 modify_width_pixels = 11 [default = 100];

  optional string sqlite3_database = 20 [default = "/tmp/liaison_commander_autosave.db"];  
  optional int32 database_pool_size = 21 [default=10];
  optional int32 database_view_height = 5  [default=400];

  enum Column
  {
    @PROTOBUF_ALLOW_ALIAS@
    COLUMN_COMMENT = 0;
    COLUMN_NAME = 1;
    COLUMN_IP = 2;
    COLUMN_TIME = 3;
    COLUMN_LAST_ACK = 4;
    COLUMN_MAX = 4;
  }
  
  message DatabaseColumnWidthPixels
  {
    optional int32 comment_width = 1 [default=180];
    optional int32 name_width = 2 [default=180];
    optional int32 ip_width = 3 [default=120];
    optional int32 time_width = 4 [default=120];
    optional int32 last_ack_width = 5 [default=120];
  }
  optional DatabaseColumnWidthPixels database_width = 22;

  message ModalDialogDimensions
  {
    optional int32 width = 1 [default = 800];
    optional int32 height = 2 [default = 200];
  }
  optional ModalDialogDimensions modal_dimensions = 23;
  
  repeated string subscription = 30;

  optional string time_source_var = 40;

  optional string network_ack_var = 41 [default = "ACOMMS_NETWORK_ACK"];
}

message NetworkAckSet
{
  optional uint64 newest_ack_time = 1;
  repeated goby.acomms.protobuf.NetworkAck ack = 2;
}

message MOOSScopeConfig
{
  repeated string subscription = 1;

  enum Column
  {
    @PROTOBUF_ALLOW_ALIAS@
    COLUMN_KEY = 0;
    COLUMN_TYPE = 1;
    COLUMN_VALUE = 2;
    COLUMN_TIME = 3;
    COLUMN_COMMUNITY = 4;
    COLUMN_SOURCE = 5;
    COLUMN_SOURCE_AUX = 6;
    COLUMN_MAX = 6;
    
  }
  
  message ColumnWidthPixels
  {
    optional int32 key_width = 1 [default=150];
    optional int32 type_width = 2 [default=60];
    optional int32 value_width = 3 [default=200];
    optional int32 time_width = 4 [default=150];
    optional int32 community_width = 5 [default=80];
    optional int32 source_width = 6 [default=80];
    optional int32 source_aux_width = 7 [default=120];
  }
  optional ColumnWidthPixels column_width = 2;
  optional Column sort_by_column = 3 [default=COLUMN_KEY];
  optional bool sort_ascending = 4 [default = true];
  optional int32 scope_height = 5  [default=400];

  optional Column regex_filter_column = 6 [default=COLUMN_KEY];
  optional string regex_filter_expression = 7 [default=".*"];

  
  message HistoryConfig
  {
    required string key = 1;
    optional bool show_plot = 2 [default = false];
    optional int32 plot_width = 3 [default = 800];
    optional int32 plot_height = 4 [default = 300];
  }
  
  
  repeated HistoryConfig history = 8;

  optional bool use_binary_protobuf = 9 [default=false];
    
    
}

message GeodesyConfig
{
  required double lat_origin = 1;
  required double lon_origin = 2;
}


message AcommsConfig
{
  optional bool minimize_dccl = 1 [default = false];
  optional bool minimize_queue = 2 [default = false];
  optional bool minimize_amac = 3 [default = false];
  optional bool minimize_driver = 4 [default = false];
  
}
